File: backend/connectors/manual.py
"""Manual entry fallback connector for platforms without API access."""

from backend.connectors.base import PlatformConnector, MetricResult


class ManualConnector(PlatformConnector):
    """Fallback for platforms that cannot be auto-tracked.

    Used for: Product Hunt, Instagram, LinkedIn, TikTok, and any
    platform where we don't have API access.
    """

    platform = "manual"

    async def validate_credentials(self) -> bool:
        """No credentials needed for manual entry."""
        return True

    async def fetch_post_metrics(self, platform_post_id: str, **kwargs) -> MetricResult:
        """Manual connector cannot fetch metrics automatically."""
        return MetricResult(fetched_via="manual")

    async def resolve_post_id(self, url: str) -> str | None:
        """For manual connectors, the URL itself serves as the identifier."""
        return url if url else None

    def is_api_trackable(self) -> bool:
        return False

File: src-tauri/src/server/connectors/manual.rs
use async_trait::async_trait;
use crate::server::connectors::{PlatformConnector, MetricResult};

pub struct ManualConnector {
    platform_name: String,
}

impl ManualConnector {
    pub fn new(platform: &str) -> Self {
        Self { platform_name: platform.to_string() }
    }
}

#[async_trait]
impl PlatformConnector for ManualConnector {
    fn platform(&self) -> &str { &self.platform_name }
    async fn validate_credentials(&self) -> bool { true }

    async fn fetch_post_metrics(&self, _platform_post_id: &str) -> anyhow::Result<MetricResult> {
        Ok(MetricResult { fetched_via: "manual".into(), ..Default::default() })
    }

    fn resolve_post_id(&self, url: &str) -> Option<String> {
        Some(url.to_string())
    }

    fn is_api_trackable(&self) -> bool { false }
}
